File: ios/zing_sdk_initializer/Sources/zing_sdk_initializer/AuthStateStreamHandler.swift
import Flutter
import Combine
import ZingCoachSDK

final class AuthStateStreamHandler: NSObject, FlutterStreamHandler {
    private let sdk: ZingSDK
    private var cancellable: AnyCancellable?

    init(sdk: ZingSDK) {
        self.sdk = sdk
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        cancellable?.cancel()
        cancellable = sdk.loginStatePublisher
            .receive(on: DispatchQueue.main)
            .sink { state in
                events(state.toFlutter())
            }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        cancellable?.cancel()
        cancellable = nil
        return nil
    }
}

private extension ZingSDK.LoginState {
    func toFlutter() -> [String: String] {
        switch self {
        case .loggedOut:
            return ["state": "loggedOut"]
        case .inProgress:
            return ["state": "inProgress"]
        case .loggedIn:
            return ["state": "authenticated"]
        }
    }
}

File: android/src/main/kotlin/com/example/zing_sdk_initializer/AuthStateStreamHandler.kt
package com.example.zing_sdk_initializer

import coach.zing.fitness.coach.SdkAuthState
import coach.zing.fitness.coach.ZingSdk
import io.flutter.plugin.common.EventChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

class AuthStateStreamHandler(
    private val scope: CoroutineScope
) : EventChannel.StreamHandler {

    private var job: Job? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        job = scope.launch {
            ZingSdk.authState.collect { state ->
                val map = when (state) {
                    is SdkAuthState.LoggedOut -> mapOf("state" to "loggedOut")
                    is SdkAuthState.InProgress -> mapOf("state" to "inProgress")
                    is SdkAuthState.Authenticated -> mapOf("state" to "authenticated")
                    else -> return@collect
                }
                events.success(map)
            }
        }
    }

    override fun onCancel(arguments: Any?) {
        job?.cancel()
        job = null
    }
}
